File: Ahorn/entities/MonumentPressureplate.jl
module EmHelperMonumentpressureplate

using ..Ahorn, Maple

@mapdef Entity "EmHelper/Monumentpressureplate" Monumentpressureplate(x::Integer, y::Integer, pattern::Integer=0, onetime::Bool=false, color::String="82d9ff", mute::Bool=false, isButton::Bool=false, disable::Bool=false)

const placements = Ahorn.PlacementDict(
   "Monument Pressureplate (EmHelper)" => Ahorn.EntityPlacement(
      Monumentpressureplate
    )
)
function getColor(color)
    if haskey(Ahorn.XNAColors.colors, color)
        return Ahorn.XNAColors.colors[color]

    else
        try
            return ((Ahorn.argb32ToRGBATuple(parse(Int, replace(color, "#" => ""), base=16))[1:3] ./ 255)..., 1.0)

        catch

        end
    end

    return (1.0, 1.0, 1.0, 1.0)
end

function Ahorn.selection(entity::Monumentpressureplate)
    x, y = Ahorn.position(entity)

    return Ahorn.Rectangle(x-6, y-4, 12, 4)
end


function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::Monumentpressureplate, room::Maple.Room)
    sprite = "objects/monumentpressureplate/solid"
    entitypattern = Int(get(entity.data, "pattern", 0))

    entitycolor = String(get(entity.data, "color", "82d9ff"))
    entitytint = getColor(entitycolor)

    Ahorn.drawSprite(ctx, sprite*string(entitypattern), 0, -8, tint=entitytint)
end
end

File: Loenn/entities/Monumentpressureplate.lua
local drawableSprite = require("structs.drawable_sprite")
local xnaColors = require("consts.xna_colors")
local utils = require("utils")
local Monumentpressureplate = {}

Monumentpressureplate.name = "EmHelper/Monumentpressureplate"
Monumentpressureplate.depth = 2000

local patterns = {
    "objects/monumentpressureplate/solid0",
    "objects/monumentpressureplate/solid1",
    "objects/monumentpressureplate/solid2",
    "objects/monumentpressureplate/solid3"
}

Monumentpressureplate.fieldInformation = {
    color = {
        fieldType = "color",
        allowXNAColors = true,
    }
}

Monumentpressureplate.placements = {
    
    name = "Monument Pressure Plate [EmHelper]",
    data = {
        pattern = 0,
        onetime = false,
        color = "LightSkyBlue",
        mute = false,
        isButton = false,
        disable = false
        }
    
}

function Monumentpressureplate.sprite(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local pattern = (entity.pattern % 4) or 0

    local PressurePlateSprite = drawableSprite.fromTexture(patterns[pattern+1], entity)
    local eSprite = drawableSprite.fromTexture(eTexture, entity)

    PressurePlateSprite:setJustification(1, 1)
    PressurePlateSprite:addPosition(8, 0)

    return PressurePlateSprite
end

function Monumentpressureplate.rectangle(room, entity)
    return utils.rectangle(entity.x-6, entity.y-4, 12, 4)
end

return Monumentpressureplate